### PR TITLE
Support IIe aux slot: 80-col(1KiB) card via generic read ptrs

### PR DESCRIPTION
**Pull request by tomcw**
PR for #1341

### CODE REVIEW

**Comment by tomcw** [comment]
Running the built-in benchmark 3 times: (Release build)

-aux ext80
Pure CPU MHz:	578.0 (full-speed)
Pure CPU MHz:	605.3 (full-speed)
Pure CPU MHz:	602.7 (full-speed)

-aux empty
Pure CPU MHz:	445.6 (full-speed)
Pure CPU MHz:	418.0 (full-speed)
Pure CPU MHz:	451.6 (full-speed)

So the old CPU emulation is 34% faster than the new alternate version (605/451=1.34).

Complete benchmark data: [benchmark.txt](https://github.com/user-attachments/files/19412396/benchmark.txt)

**Comment by tomcw** [comment]
@Michaelangel007 - a large PR. I don't expect a detailed code review, but mainly just giving you a heads-up about it.

I wrote a few high-level notes about it here: https://github.com/AppleWin/AppleWin/issues/1353#issuecomment-2746562623

TL;DR - this is to better support the //e's aux slot when it's empty or has the 80-col(1KiB) card.

When the aux slot has 80-col(64KiB) or Ramworks III cards (or it's not a IIe being emulated), then the emulation reverts to the original emulation code. So in terms of any regressions I may have added, then these should just be constrained to these edge case config of aux = empty/80-col(1KiB). The only other area impacted is the debugger, where I've changed all the direct `mem` accesses to go via `ReadByteFromMemory()`... and how to test all of these?!

For the future, one direction this could go in (not for this PR) is we drop the `mem` read cache (so no more `UpdatePaging()` and `memdirty[]`), which simplifies the core CPU emulation code... at the expense of some performance.

btw, on the subject of performance, we really need a x86-64 build, which will give us a big perf boost (mainly due to all the extra registers available).

**Comment by Michaelangel007** [comment]
> btw, on the subject of performance, we really need a x86-64 build, which will give us a big perf boost (mainly due to all the extra registers available).

Yes, a native x64 build would be very nice!

I was curious what this would entail so I tried:

1. Right-click on the `AppleWin` project
2. Properties
3. `[Configuration Manager]`
4. _Active Solution Platform_ -> `<New...>`
5. _Type of select the new platform_ -> `x64`
6. `[OK]`

It created Debug x64 and Release x64 for all projects which is nice.

Surprisingly it worked!  I guess all the clean-up work everyone has done over the years has paid off -- I was expecting the usual 32-bit -> 64-bit tranistion problems.  I would feel better though if we did a "full pass" of testing things out to make sure everything is good.

We **probably should have the Debugger's `CmdMOTD()` display the platform** so we easily tell if we are running the `x86` or `x64` version.

I took a look at the `.vcxproj` -- it has new lines along:

```
    <ProjectConfiguration Include="Release|x64">
      <Configuration>Release</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
```

and

```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
    <ConfigurationType>Application</ConfigurationType>
    <UseDebugLibraries>false</UseDebugLibraries>
    <PlatformToolset>v143</PlatformToolset>
    <WholeProgramOptimization>false</WholeProgramOptimization>
    <CharacterSet>MultiByte</CharacterSet>
  </PropertyGroup>
```

Looks like we might want to play with some of the optimization settings?

**Comment by Michaelangel007** [comment]
> Running the built-in benchmark 3 times: (Release build)
> -aux ext80 Pure CPU MHz: 578.0 (full-speed) Pure CPU MHz: 605.3 (full-speed) Pure CPU MHz: 602.7 (full-speed)
> -aux empty Pure CPU MHz: 445.6 (full-speed) Pure CPU MHz: 418.0 (full-speed) Pure CPU MHz: 451.6 (full-speed)
> 
> So the old CPU emulation is 34% faster than the new alternate version (605/451=1.34).

Can you walk-me through what I need to do to benchmark this?  Are you just running with the command line, and then using Configuration > `[Benchmark Emulator]` ?

Q. Can we "select/copy" the test of that benchmark summary dialog box?
A. Ooh, yes we can! **Looks like we need to update the help under `Configuration Settings`** with something along the lines of :  _One can press `Ctrl`-`C` at this dialog to put the results on the clipboard._

This is the older 1.30.20.0 for my system:

```
---------------------------
Benchmarks
---------------------------
Pure Video FPS:	3849 hires, 3853 text
Pure CPU MHz:	17.2 (video update)
Pure CPU MHz:	342.9 (full-speed)

EXPECTED AVERAGE VIDEO GAME
PERFORMANCE: 925 FPS
---------------------------
OK   
---------------------------
```

**Comment by Michaelangel007** [comment]
> I don't expect a detailed code review, but mainly just giving you a heads-up about it.

Now that my [Jet Color Mapping Comparison](https://www.shadertoy.com/view/tc2SDw) shader is done I can take a look at this PR this week.  I'll give a quick read-through one each of the commits and give a thumbs up or ask-for-clarification.

**Comment by tomcw** [comment]
>This is the older 1.30.20.0 for my system:
>`Pure CPU MHz:	342.9 (full-speed)`

So how did the 64-bit version compare?

(btw, #1392 to track this 64-bit work)

**Comment by Michaelangel007** [comment]
> So how did the 64-bit version compare?

[Results here](https://github.com/AppleWin/AppleWin/issues/1392#issuecomment-2749370662)

[Summary here](https://github.com/AppleWin/AppleWin/issues/1392#issuecomment-2752347360)

File: source/CPU.cpp [hunk 1]
```diff
@@ -777,7 +816,7 @@ void CpuReset()
 
 //===========================================================================
 
-void CpuSetupBenchmark ()
```
Comment (Michaelangel007) [inline source/CPU.cpp:780]: I was leaving a space after the function definition and before the opening parenthesis to make it super fast to find via a text search.

Shame to lose this.  Do we need to update the coding standard?

File: source/CPU/cpu6502.h [hunk 2]
```diff
@@ -21,6 +21,55 @@ along with AppleWin; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
 
+// For regular or alternate (slow-path) CPU emulation
```
Comment (Michaelangel007) [inline source/CPU/cpu6502.h:24]: I find this easier to read:
```
// For regular or alternate (slow-path) CPU emulation
#ifndef CPU_ALT
//#define READ         _READ
//#define WRITE(value) _WRITE(value)
#define JSR         _JSR
#define POP         _POP
#define PUSH(value) _PUSH(value)
#define ABS         _ABS
#define IABSX       _IABSX
#define ABSX_CONST _ABSX_CONST
#define ABSX_OPT   _ABSX_OPT
#define ABSY_CONST _ABSY_CONST
#define ABSY_OPT   _ABSY_OPT
#define IABS_CMOS  _IABS_CMOS
#define IABS_NMOS  _IABS_NMOS
#define INDX       _INDX
#define INDX       _INDX
#define INDY_CONST _INDY_CONST
#define INDY_OPT   _INDY_OPT
#define IZPG       _IZPG
#define REL        _REL
#define ZPG        _ZPG
#define ZPGX       _ZPGX
#define ZPGY       _ZPGY
#else
//#define READ         _READ_ALT
//#define WRITE(value) _WRITE_ALT(value)
#define JSR         _JSR_ALT
#define POP         _POP_ALT
#define PUSH(value) _PUSH_ALT(value)
#define ABS         _ABS_ALT
#define IABSX       _IABSX_ALT
#define ABSX_CONST  _ABSX_CONST_ALT
#define ABSX_OPT    _ABSX_OPT_ALT
#define ABSY_CONST  _ABSY_CONST_ALT
#define ABSY_OPT    _ABSY_OPT_ALT
#define IABS_CMOS   _IABS_CMOS_ALT
#define IABS_NMOS   _IABS_NMOS_ALT
#define INDX        _INDX_ALT
#define INDX        _INDX_ALT
#define INDY_CONST  _INDY_CONST_ALT
#define INDY_OPT    _INDY_OPT_ALT
#define IZPG        _IZPG_ALT
#define REL         _REL_ALT
#define ZPG         _ZPG_ALT
#define ZPGX        _ZPGX_ALT
#define ZPGY        _ZPGY_ALT
#endif
```

File: source/CPU/cpu6502.h [hunk 3]
```diff
@@ -21,6 +21,55 @@ along with AppleWin; if not, write to the Free Software
...
+#define INDX _INDX
+#define INDY_CONST _INDY_CONST
+#define INDY_OPT _INDY_OPT
+#define IZPG _IZPG
+#define REL _REL
+#define ZPG _ZPG
+#define ZPGX _ZPGX
+#define ZPGY _ZPGY
+#else
+//#define READ _READ_ALT
```
Comment (Michaelangel007) [inline source/CPU/cpu6502.h]: Could we get a (short) comment for why `READ` and `WRITE` are commented out?  Deprecated? Redundant?

**Review by Michaelangel007 — COMMENTED** [review]
Maybe cleanup the non-descript `loop` variable names?  Does it refer to?

* _byte offsets_? 
* _page numbers_?

**Comment by Michaelangel007** [comment]
**TL:DR;** Reviewed 15 / 29 = 51%.

Review summary so I can track my progress due to the many files being changed:

* [x] [0f44689](https://github.com/AppleWin/AppleWin/pull/1391/commits/0f446893584135a7ff7067d6ef3115d6d45d463b)
* [x] [f71bd55](https://github.com/AppleWin/AppleWin/pull/1391/commits/f71bd55f0497fcc531cbdf81348457595b2633af) 
* [x] [dcd4e52](https://github.com/AppleWin/AppleWin/pull/1391/commits/dcd4e5233a7094c667e02cd6ee573395d9e6fc56)
* [x] [086197f](https://github.com/AppleWin/AppleWin/pull/1391/commits/086197f4102d4a03d718f8a5a24f2d5a3497dfa9)
* [x] [e4968b1](https://github.com/AppleWin/AppleWin/pull/1391/commits/e4968b11cc3dbf0aba683263ea4f9a65e60e7996)
* [x] [9dcc9a3](https://github.com/AppleWin/AppleWin/pull/1391/commits/9dcc9a376ae65a2bcb1717f2e372dbf233ef8cfb)
* [x] [a99c795](https://github.com/AppleWin/AppleWin/pull/1391/commits/a99c7953ffb98abdbe8ca1c16a0cfb949b38a846)
* [x] [048a15b](https://github.com/AppleWin/AppleWin/pull/1391/commits/048a15b5032f47e7f8726c7dc8bc122c6d10ed94)
* [x] [3cfd2fd](https://github.com/AppleWin/AppleWin/pull/1391/commits/3cfd2fd4d79a23c863c0893415c21465613c7514)
* [x] [b26fc94](https://github.com/AppleWin/AppleWin/pull/1391/commits/b26fc9413e60133dc8b61ab3a702589a782a9ac7)
* [x] [9703a79](https://github.com/AppleWin/AppleWin/pull/1391/commits/9703a791f1e3fd4c022f062cbb5d27abd96dcdaa)
* [x] [2f7e271](https://github.com/AppleWin/AppleWin/pull/1391/commits/2f7e27170c95930bd8a1f1bcdc4784679d280d7f)
* [x] [f078e82](https://github.com/AppleWin/AppleWin/pull/1391/commits/f078e82c1b4d35104eee72ebb208a889a2e4b32a)
* [x] [56685bc](https://github.com/AppleWin/AppleWin/pull/1391/commits/56685bcd2888195ba8fb901a7b8a409cefe99264)
* [x] [61a0e99](https://github.com/AppleWin/AppleWin/pull/1391/commits/61a0e99312abd2b487a701e88db79722b556f720)
* [ ] [e985a77](https://github.com/AppleWin/AppleWin/pull/1391/commits/e985a77aaa70a54a9222fea0c35001af95775860)
* [ ] [c34306d](https://github.com/AppleWin/AppleWin/pull/1391/commits/c34306d702d9d1cf3c6bea682e3fc0a9943d9dd8)
* [ ] [acd8099](https://github.com/AppleWin/AppleWin/pull/1391/commits/acd80995e28cd8af6feae30daae4f92209a53025)
* [ ] [789b573](https://github.com/AppleWin/AppleWin/pull/1391/commits/789b5733414b2d42ce0db5279c0b0562a7fbb054)
* [ ] [36c916d](https://github.com/AppleWin/AppleWin/pull/1391/commits/36c916d8177bdd6cbfbbc60d1d0c82b92c962b4e)
* [ ] [a118ad7](https://github.com/AppleWin/AppleWin/pull/1391/commits/a118ad747bb9f4b85c64e135c13d9609503b1fb2)
* [ ] [01f1536](https://github.com/AppleWin/AppleWin/pull/1391/commits/01f1536720f1935e5dbd9d3f638d8fc6f9439a54)
* [ ] [f729ca7](https://github.com/AppleWin/AppleWin/pull/1391/commits/f729ca72a84a1ef4f36281e772305c7f10ee3c1d)
* [ ] [f1266b4](https://github.com/AppleWin/AppleWin/pull/1391/commits/f1266b445a489764ec9a3fa080b69b627caacd31)
* [ ] [e8a5994](https://github.com/AppleWin/AppleWin/pull/1391/commits/e8a5994f41fb2787459f173f7c2ca2bf3e6726d2)
* [ ] [3300438](https://github.com/AppleWin/AppleWin/pull/1391/commits/33004388eaaa1bd21ae07ca3cc13b6ead59ea8a3)
* [ ] [b36d6a7](https://github.com/AppleWin/AppleWin/pull/1391/commits/b36d6a7c413bfc524756642bd5f19661beb8606b)
* [ ] [84a7a1e](https://github.com/AppleWin/AppleWin/pull/1391/commits/84a7a1ecdef94200e0595439e901d980f0a1d50f)
* [ ] [8c45563](https://github.com/AppleWin/AppleWin/pull/1391/commits/8c455637e56d1a29a56b959bf0a351abb2868cd9)

Whew!

File: source/CPU.cpp (hunk 1 above)
Comment (tomcw) [reply source/CPU.cpp:780]: I think that "style" ship has mostly sailed in the _main body_ of source code.
From Visual Studio, F12=Go to definition; CTRL+F12=Go to declaration; SHIFT+F12=Find all references

But in the debugger body of source code, this style still dominates, and I try to adhere to it to be consistent with all the Debugger source. And in the CodingConvention.txt, we already have this documented - see "2.6.1.1: Debugger function style".

**Comment by Michaelangel007** [comment]
@tomcw I took a quick look though all the commits. (Whew!)

Overall this patch looks great!

Thanks for taking the time to cleanup some of the magic-numbers scattered throughout.

I left some comments for the things that will help the codebase maintainability and readability over the long term.  Nothing big, just small stuff here and there.

* Would like to see all the CPU addressing modes aligned.
* I'm fine with leaving the 6502 constants referencing memory; save that for a smaller future PR for `MemoryDefs.h`
* I'm _fine_ with leaving `loop` as-is; save that for a smaller future PR.  Only reason to touch now is that it still may be fresh in your head? i.e. https://github.com/AppleWin/AppleWin/pull/1391/commits/f71bd55f0497fcc531cbdf81348457595b2633af
* I'm fine with also leaving the benchmark magic numbers as-is for; save that for a smaller future PR cleanup.

Personally, all that memory caching stuff (`mem` / `memshadow`) gives me a headache -- for some reason it has never really "clicked" with me. This PR / cleanup greatly manage this complexity.

Did I mention that `Cs00..CsFF` is a beautiful, succinct comment. ;-)

**Comment by tomcw** [comment]
Thanks for taking the time to go through all of these changes - there are certainly a lot!

I'll take a look at all the points you've raised over the coming days.

I've also been testing the Debugger changes (eg. CmdMemoryLoad/Move/Save), and these all continue to behave correctly. I plan to do some coverage testing of all the other changes, so as to increase my confidence.

>Did I mention that `Cs00..CsFF` is a beautiful, succinct comment. ;-)

Yeh, I'll improve that! :-)

**Comment by tomcw** [comment]
@Michaelangel007 - 3 commits to address your review comments (note in the 1st commit I didn't quite get what you asked, so see the 2nd commit at the same time!)

**Review by Michaelangel007 — APPROVED** [review]
👍

File: source/Debugger/Debugger_Assembler.cpp [hunk 4]
```diff
@@ -477,8 +477,8 @@ WORD _6502_GetStackReturnAddress ()
 WORD _6502_PeekStackReturnAddress (WORD & nStack)
 {
 	WORD   nAddress;
-	       nAddress  = ((unsigned) *(LPBYTE)(mem + 0x100 + (nStack & 0xFF))     ); nStack++;
-	       nAddress += ((unsigned) *(LPBYTE)(mem + 0x100 + (nStack & 0xFF)) << 8);
+	       nAddress  =  ReadByteFromMemory(_6502_STACK_BEGIN + (nStack & 0xFF)); nStack++;
+	       nAddress += (ReadByteFromMemory(_6502_STACK_BEGIN + (nStack & 0xFF)) << 8);
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Debugger/Debugger_Assembler.cpp:481]: I've consolidated DebugDefs.h into MemoryDefs.h (and deleted DebugDefs.h).
I've split into 2 enums: `_6502` prefixed and `APPLE` prefixed, depending on whether it's 6502 or Apple definitions.

File: source/Memory.h [hunk 5]
```diff
@@ -114,3 +114,6 @@ void	SetRegistryAuxNumberOfBanks(void);
 
 LPBYTE GetCxRomPeripheral(void);
 bool GetIsMemCacheValid(void);
+BYTE ReadByteFromMemory(uint16_t addr);
+void WriteByteToMemory(uint16_t addr, uint8_t data);
+void CopyBytesFromMemoryPage(uint8_t* pDst, uint16_t srcAddr, size_t size);
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.h:120]: I don't think we need the comment - as we both know to retain this `memcpy()` arg order, and we'll both be involved in reviewing any PR if someone were to try to switch the arg order.

Long-term, `Memory.cpp` should become a class, then we don't need any `MemXXX*()` prefixes.

File: source/Memory.cpp [hunk 6]
```diff
@@ -2522,21 +2531,25 @@ bool IsIIeWithoutAuxMem(void)
...
-
 		// IF THE EMULATED PROGRAM HAS JUST UPDATED THE MEMORY WRITE MODE AND IS
 		// ABOUT TO UPDATE THE MEMORY READ MODE, HOLD OFF ON ANY PROCESSING UNTIL
 		// IT DOES SO.
 		//
 		// NB. A 6502 interrupt occurring between these memory write & read updates could lead to incorrect behaviour.
 		// - although any data-race is probably a bug in the 6502 code too.
 		if ((address >= 4) && (address <= 5) &&									// Now:  RAMWRTOFF or RAMWRTON
-			((*(LPDWORD)(mem+programcounter) & 0x00FFFEFF) == 0x00C0028D))		// Next: STA $C002(RAMRDOFF) or STA $C003(RAMRDON)
+			((ReadUINT24FromMemory(programcounter) & 0x00FFFEFF) == 0x00C0028D))		// Next: STA $C002(RAMRDOFF) or STA $C003(RAMRDON)
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.cpp:2616]: No it's not public (hence the `static`) - in the future, this function could become a lamda and be moved into to the body of the `MemOptimizeForModeChanging()`.

File: source/CPU.cpp [hunk 7]
```diff
@@ -512,7 +511,7 @@ static __forceinline bool IRQ(ULONG& uExecutedCycles, BOOL& flagc, BOOL& flagn,
 #define HEATMAP_X(address)
 
 // 6502 & no debugger
-#define READ _READ_WITH_IO_F8xx
+#define READ(addr) _READ_WITH_IO_F8xx(addr)
 #define WRITE(value) _WRITE_WITH_IO_F8xx(value)
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/CPU.cpp:515]: Done.

File: source/CPU.h [hunk 8]
```diff
@@ -25,6 +25,10 @@ enum {
 	AF_CARRY = 0x01
 };
 
+#define VECTOR_INTERRUPT 0xFFFE	/* IRQ & BRK */
+#define VECTOR_RESET 0xFFFC
+#define VECTOR_NMI 0xFFFA
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/CPU.h]: Yes... and moved into MemoryDefs.h.

File: source/Windows/Win32Frame.cpp [hunk 9]
```diff
@@ -206,17 +208,17 @@ void Win32Frame::Benchmark(void)
 	// SIMULATE THE ACTIVITY OF AN AVERAGE GAME
 	uint32_t totalhiresfps = 0;
 	video.SetVideoMode(VF_HIRES);
-	memset(mem + 0x2000, 0x14, 0x2000);
+	memset(pMemMain + 0x2000, 0x14, 0x2000);
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Windows/Win32Frame.cpp]: Added to `MemoryDefs.h` and applied to `Benchmark()`.

File: source/DiskImageHelper.cpp [hunk 10]
```diff
@@ -1038,7 +1038,7 @@ class CAplImage : public CImageBase
 			return false;
 		}
 
-		ReadFile(ptr->hFile, mem+address, length, &bytesread, NULL);
+		ReadFile(ptr->hFile, MemGetMainPtr(address), length, &bytesread, NULL);
 		int loop = 192;
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/DiskImageHelper.cpp]: Done

File: source/Memory.cpp [hunk 11]
```diff
@@ -1067,7 +1072,7 @@ void RegisterIoHandler(UINT uSlot, iofunction IOReadC0, iofunction IOWriteC0, io
 	if (IOReadCx == NULL)	IOReadCx = IO_Cxxx;
 	if (IOWriteCx == NULL)	IOWriteCx = IO_Cxxx;
 
-	for (UINT i=0; i<16; i++)
+	for (UINT i=0; i<16; i++)	// Cs00..CsFF
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.cpp]: Improved this comment.

File: source/CPU/cpu_general.inl [hunk 12]
```diff
@@ -35,7 +35,8 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 #undef AF_TO_EF
 #undef EF_TO_AF
 
-
+#define ZERO_PAGE 0x00
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/CPU/cpu_general.inl]: Yes, good point. Done.

File: source/Memory.cpp [hunk 13]
```diff
@@ -1617,6 +1617,9 @@ LPBYTE MemGetMainPtr(const WORD offset)
 
 static void BackMainImage(void)
 {
+	if (!g_isMemCacheValid)
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.cpp:1706]: Done

File: source/Memory.cpp [hunk 14]
```diff
@@ -1525,7 +1525,7 @@ static LPBYTE MemGetPtrBANK1(const WORD offset, const LPBYTE pMemBase)
 
 	// NB. This works for memaux when set to any RWpages[] value, ie. RamWork III "just works"
 	const BYTE bank1page = (offset >> 8) & 0xF;
-	return (memshadow[0xD0+bank1page] == pMemBase+(0xC0+bank1page)*256)
+	return (g_isMemCacheValid && (memshadow[0xD0+bank1page] == pMemBase+(0xC0+bank1page)*256))
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.cpp]: Done

File: source/Memory.h [hunk 15]
```diff
@@ -43,9 +43,9 @@ typedef BYTE (__stdcall *iofunction)(WORD nPC, WORD nAddr, BYTE nWriteFlag, BYTE
 
 extern iofunction IORead[256];
 extern iofunction IOWrite[256];
+extern LPBYTE     memshadow[0x100];
 extern LPBYTE     memwrite[0x100];
 extern BYTE       memreadPageType[0x100];
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Memory.h:48]: Replaced with `_6502_NUM_PAGES`

File: source/MouseInterface.cpp [hunk 16]
```diff
@@ -235,10 +235,11 @@ void CMouseInterface::SetSlotRom()
 	if (pCxRomPeripheral == NULL)
 		return;
 
-	UINT uOffset = (m_by6821B << 7) & 0x0700;
-	memcpy(pCxRomPeripheral+m_slot*256, m_pSlotRom+uOffset, 256);
+	const UINT offset = (m_by6821B << 7) & 0x0700;
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/MouseInterface.cpp:239]: I've added a comment, as the VIA chip's usage by the Mousecard is implementation specific (ie. not defined on the VIA's datasheet).

**Comment by tomcw** [comment]
Thanks @Michaelangel007 for all those extra review comments. I've addressed them all.

I'll do a bit more testing, and then merge into the main line.

**Review by Michaelangel007 — COMMENTED** [review]
Looks good!

File: source/CPU/cpu_general.inl [hunk 17]
```diff
@@ -83,7 +80,7 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 					: MemReadFloatingBus(uExecutedCycles)						\
 		)
 #define _READ_WITH_IO_F8xx(addr) (									/* GH#827 */\
-			((addr & 0xF000) == 0xC000)											\
+			((addr & 0xF000) == APPLE_IO_BEGIN)									\
```
Comment (Michaelangel007) [inline source/CPU/cpu_general.inl:83]: I _really_ like the usage of `_6502_*` for generic 6502 stuff and `APPLE_IO_*` for stuff specifically for the Apple 2.  I didn't think of that. VERY nice.

**Review by Michaelangel007 — COMMENTED** [review]
👍

**Review by Michaelangel007 — COMMENTED** [review]
👍

**Review by Michaelangel007 — COMMENTED** [review]
👍 I'm happy with these for commit [e74624c](https://github.com/AppleWin/AppleWin/pull/1391/commits/e74624ca451a4171dae178e3066d3604f2474c47)

Minor question: Q. Do we want to use the `APPLE_IO_` prefix for `TEXT_PAGE1_BEGIN`, etc.?

`APPLE_IO_TEXT1_PAGE1_BEGIN`:

* Pro:  is more descriptive so that the reader knows this is specific to the Apple 2.
* Con: Is getting to be TOO verbose/long?

`TEXT_PAGE1_BEGIN`:

* Pro: Compact, yet descriptive enough.
* Con: Assumes reader knows this is Apple 2 specific.

I'm find with `TEXT_PAGE1_BEGIN` since for the entire project the context is specifically the Apple 2 but was curious if there was any benefit to be more (slightly more) consistent with `APPLE_IO` ?

**Review by Michaelangel007 — COMMENTED** [review]
👍 for commit [24207ee](https://github.com/AppleWin/AppleWin/pull/1391/commits/24207eea751555bf305c6747dbc40fc1180932a3)

**Review by Michaelangel007 — COMMENTED** [review]
👍 for commit [https://github.com/AppleWin/AppleWin/pull/1391/commits/55a62c6e237d0fbf36de50f3b822ba70801882a5](https://github.com/AppleWin/AppleWin/pull/1391/commits/ce237d0fbf36de50f3b822ba70801882a5)

**Comment by tomcw** [comment]
>Minor question: Q. Do we want to use the `APPLE_IO_` prefix for `TEXT_PAGE1_BEGIN`, etc.?

Yeah, maybe... I like your idea - all these enums would have a common prefix.
Perhaps now `APPLE_IO_`, since they aren't all "IO" memory, so maybe `APPLE_` instead?

Anyway, I think park that idea for now, and we can tackle it in a dedicated PR later on.

**Comment by Michaelangel007** [comment]
> so maybe `APPLE_` instead?

That's even better!

> Anyway, I think park that idea for now, and we can tackle it in a dedicated PR later on.

Agreed.

**Comment by tomcw** [comment]
@Michaelangel007 - FYI, I've gone ahead and merged this PR into the main line.
